pkg/scheduler/util/utils.go

```diff
@@ -115,13 +126,48 @@ func PatchPodStatus(cs kubernetes.Interface, old *v1.Pod, newStatus *v1.PodStatu
 		return nil
 	}
 
-	_, err = cs.CoreV1().Pods(old.Namespace).Patch(context.TODO(), old.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{}, "status")
-	return err
+	backoff := wait.Backoff{
+		Duration: retryBackoffInitialDuration,
+		Factor:   retryBackoffFactor,
+		Jitter:   retryBackoffJitter,
+		Steps:    retryBackoffSteps,
+	}
+
+	patchFn := func() error {
+		_, err := cs.CoreV1().Pods(old.Namespace).Patch(context.Background(), old.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+		return err
+	}
+
+	isRetriableFn := func(err error) bool {
+		// Check if the error is returned from kube-apiserver.
+		// If so, check retryAfterSeconds to know if kube-apiserver asked us to retry or not.
+		typedErr, ok := err.(*errors.StatusError)
+		if !ok {
+			// unknown error. e.g. api-server is unreachable
+			// let's retry it and see how it goes.
+			return true
+		}
+
+		var retryAfterSeconds int32
+		if typedErr.Status().Details != nil {
+			retryAfterSeconds = typedErr.Status().Details.RetryAfterSeconds
+		}
+		if retryAfterSeconds == 0 {
+			// we shouldn't retry since this means kube-apiserver rejected our request and didn't ask us to retry.
+			return false
+		}
+		// Otherwise we can retry after retryAfterSeconds.
+		klog.ErrorS(err, "Server rejected Pod patch (may retry after sleeping)", "pod", klog.KObj(old))
+		time.Sleep(time.Duration(retryAfterSeconds))
```

Comment:
you could actually use the `backoff.Duration` to find the diff. The field is updated every time.

Reply:
> you could actually use the backoff.Duration to find the diff. The field is updated every time.

I guessed so too, but it doesn't seem to be updated every time since `wait.ExponentialBackoff` and `` receives non-pointer `wait.Backoff` not pointer `*wait.Backoff`...

https://github.com/kubernetes/kubernetes/blob/e84f62e6955b5d45f15cc670e314523f97bb976e/staging/src/k8s.io/client-go/util/retry/util.go#L48
https://github.com/kubernetes/kubernetes/blob/e84f62e6955b5d45f15cc670e314523f97bb976e/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L419

Reply:
oh I missed that :(

Do you have any idea how long usually `retryAfterSeconds` is?

Reply:
> Do you have any idea how long usually retryAfterSeconds is?

It depends on the type of error:

```
GanerateNameConflict→1s
TooManyRequests→10s
ServerTimeout→2s
Timeout→10s
```

(investigated by searching the usage of `NewXXXXX` in [k8s.io/apimachinery/pkg/api/errors/errors.go](https://github.com/kubernetes/kubernetes/blob/fbdd0d7b4165bc5a677d45e4dc693e3260297bfa/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go).)

Reply:
Ah, that's great!

I'm happy with the backoffs you have set.

/lgtm
/approve
/hold
@ahg-g anything to add?

Reply:
If we can figure the exact failure reasons, I'd propose to restrict the retry logic to those particular error(s), like the "connection refused" one:

https://github.com/kubernetes/kubernetes/blob/93047824f7dec417bb90d89080b4550fe8b771df/staging/src/k8s.io/apimachinery/pkg/util/net/util.go#L50 can direcetly pin the error.

Reply:
> I'm curious about the exact failure reasons,

> If we can figure the exact failure reasons, I'd propose to restrict the retry logic to those particular error(s), like the "connection refused" one

At least the error we have seen is "connection refused".
https://github.com/kubernetes/kubernetes/issues/109783#issuecomment-1117670945

But, when api-server is very busy and could return `ServerTimeout` or `TooManyRequests` errors, we'll face the same issue as https://github.com/kubernetes/kubernetes/issues/109796.
So, I think it's better to retry not only for connection refused error, but also for errors which contains `retryAfterSeconds`.

Reply:
But, the current implementation retries for all unknown errors(= all errors which type is not `*errors.StatusError`), so it would make sense to change it to retry only for `connection refused` if the error is not `*errors.StatusError`.